programs/bpf_loader/src/serialization.rs

```diff
@@ -169,18 +169,25 @@ pub fn deserialize_parameters_unaligned(
             start += size_of::<u8>(); // is_signer
             start += size_of::<u8>(); // is_writable
             start += size_of::<Pubkey>(); // key
-            let _ = borrowed_account.set_lamports(LittleEndian::read_u64(
+            let lamports = LittleEndian::read_u64(
                 buffer
                     .get(start..)
                     .ok_or(InstructionError::InvalidArgument)?,
-            ));
+            );
+            if borrowed_account.get_lamports() != lamports {
+                borrowed_account.set_lamports(lamports)?;
+            }
             start += size_of::<u64>() // lamports
                 + size_of::<u64>(); // data length
-            let _ = borrowed_account.set_data(
-                buffer
-                    .get(start..start + pre_len)
-                    .ok_or(InstructionError::InvalidArgument)?,
-            );
+            let data = buffer
+                .get(start..start + pre_len)
+                .ok_or(InstructionError::InvalidArgument)?;
+            // The redundant `is_data_mutable` helps to avoid the expensive data comparison if we can
+            if borrowed_account.is_data_mutable(data.len()).is_ok()
+                || borrowed_account.get_data() != data
+            {
```

Comment:
As long as `enable_early_access_violation` is `false`, `is_data_mutable()` always returns `Ok(())` and this here evaluates to `if true`. So the behavior stays the same as it currently is.